File: solutions/go/strain/1/strain.go
package strain

func negate[T any](predicate func(T) bool) func(T) bool {
	return func(item T) bool {
		return !predicate(item)
	}
}

func Keep[T any](collection []T, predicate func(T) bool) (result []T) {
	for _, item := range collection {
		if predicate(item) {
			result = append(result, item)
		}
	}
	return
}

func Discard[T any](collection []T, predicate func(T) bool) (result []T) {
	return Keep(
		collection,
		negate(predicate),
	)
}

File: solutions/erlang/strain/1/src/strain.erl
-module(strain).

-export([keep/2, discard/2]).

-type predicate() :: fun((Element :: any()) -> boolean()).

-spec keep(Fn :: predicate(), List :: [T]) -> [T].
keep(Fn, List) ->
    keep(Fn, List, []).

-spec keep(Fn :: predicate(), List :: [T], Acc :: [T]) -> [T].
keep(Fn, [H | T], Acc) ->
    case Fn(H) of
        true ->
            keep(Fn, T, [H | Acc]);
        false ->
            keep(Fn, T, Acc)
    end;
keep(_Fn, [], Acc) ->
    lists:reverse(Acc).

-spec discard(Fn :: predicate(), List :: [T]) -> [T].
discard(Fn, List) ->
    keep(fun(E) -> not Fn(E) end, List).
